clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -169,13 +187,17 @@ public void authenticate() throws IOException {
 
         switch (saslState) {
             case SEND_APIVERSIONS_REQUEST:
-                // Always use version 0 request since brokers treat requests with schema exceptions as GSSAPI tokens
-                ApiVersionsRequest apiVersionsRequest = new ApiVersionsRequest((short) 0);
-                send(apiVersionsRequest.toSend(node, nextRequestHeader(ApiKeys.API_VERSIONS, apiVersionsRequest.version())));
-                setSaslState(SaslState.RECEIVE_APIVERSIONS_RESPONSE);
-                break;
+                if (this.apiVersionsResponse == null) {
+                    // Always use version 0 request since brokers treat requests with schema exceptions as GSSAPI tokens
+                    ApiVersionsRequest apiVersionsRequest = new ApiVersionsRequest((short) 0);
+                    send(apiVersionsRequest.toSend(node, nextRequestHeader(ApiKeys.API_VERSIONS, apiVersionsRequest.version())));
+                    setSaslState(SaslState.RECEIVE_APIVERSIONS_RESPONSE);
+                    break;
+                } else
+                    setSaslState(SaslState.RECEIVE_APIVERSIONS_RESPONSE); // fall through
             case RECEIVE_APIVERSIONS_RESPONSE:
-                ApiVersionsResponse apiVersionsResponse = (ApiVersionsResponse) receiveKafkaResponse();
+                ApiVersionsResponse apiVersionsResponse = this.apiVersionsResponse != null ? this.apiVersionsResponse
```

Comment:
Same as before, we dont need this if initial state is set for reauthentication,

Reply:
👍 Agreed, no longer needed -- the initial state for client-side re-authentication is now `PROCESS_APIVERSIONS_RESPONSE`.